### ISSUE TITLE
Title: [Bug] NaN Loss and High Test Loss During Training in Docker Environment

### ISSUE THREAD

**Issue by fengxiaocurious** [henrysky/astroNN_stars_foundation#2]
Hi,

First of all, thank you for your excellent work on this project. I'm very interested in your results and have been trying to reproduce the training process.

I've been following your instructions, using the provided Dockerfile to build the environment. However, I'm encountering a critical issue during training.

**The Problem**

The training process runs, but the loss unpredictably becomes NaN somewhere between 1000 and 3000 iterations. When this happens, all subsequent metrics also become NaN.

Here is a snippet from the output of tail -f model_torch/training_metrics.csv which shows the exact moment the error occurs:

`34.20211338996887,-1.5723792456689922,0.4086455349949585,-1.7224479913711548,0.2752747784058253,6.1254e-05
34.19414734840393,-1.5822870138732867,0.38984424768593806,-1.7068988879521687,0.3019879659016927,6.0955e-05
34.15019488334656,-1.5797081706167637,0.37393975900164966,-1.7080041964848836,0.3072091241677602,6.0656e-05
34.19312381744385,-1.5772433161050423,0.3895238193244427,-1.7230071226755779,0.30305075148741406,6.0356e-05
34.15534830093384,-1.5831174316077397,0.3815562965712328,-1.7179742654164631,0.2555319418509801,6.0055e-05
34.17791271209717,-1.5650773723235076,0.43293063781470403,-1.7099298636118572,0.32601531346638996,5.9755e-05
34.1790075302124,-1.573158984554225,0.5050146192226602,-1.7044068177541096,0.3531922698020935,5.9453e-05
34.17246627807617,-1.579274316628774,0.3995602983852913,-1.7169583638509114,0.25390419860680896,5.9152e-05
34.19146728515625,-1.5796891728351856,0.3922818894658623,-1.7275420427322388,0.24877244730790457,5.8850e-05
34.13809847831726,-1.5809247990449269,0.44300632624105474,-1.7399118741353352,0.2893330752849579,5.8548e-05
34.15339136123657,-1.579307526692577,0.3977382959819388,-1.7168396313985188,0.3214708169301351,5.8246e-05
34.17126178741455,-1.5852067542487178,0.4030089216285396,-1.7312504450480144,0.3159886399904887,5.7943e-05
34.178627729415894,-1.5831104862964016,0.37542852609493266,-1.7192491292953491,0.324970543384552,5.7640e-05
34.12706732749939,-1.5862591985998482,0.40183067413839113,-1.7138148148854573,0.29266611238320667,5.7337e-05
34.184927225112915,nan,nan,nan,nan,5.7033e-05`

This occurred without any changes to the code.

**Secondary Issue: High Test Loss**

On runs where the NaN error doesn't happen for over 2000 iterations, I've noticed that the test mse loss seems to plateau around 0.3. This is significantly higher than the loss of approximately 0.07 for your pre-trained weights. This suggests there might be a divergence in the training process even when it doesn't crash.

**Possible Cause?**

I was investigating potential causes and noticed a possible environment mismatch. In your pre-trained model files, it appears the PyTorch version used was 2.1. However, the installation requirements seem to pull a more recent version, 2.4.

I am wondering if the environment was updated recently, but perhaps other project components (like hyperparameters or certain operations) were not updated to match? A version mismatch like this could be a potential source of the training instability I'm observing.

**Questions**

i） Is this NaN loss a known issue, or have you encountered it before?

ii) Is the test loss of ~0.3 expected with the current public code, or does it also indicate a problem?

Thank you for your time and help. Any guidance you can provide would be greatly appreciated.

**Comment by henrysky** [comment]
I was on vacation and just have time to look at this issue.

The environment indeed was udpated after the paper published. But as far as I remember, I did not actually use the docker enviroment to train my model. The model was trained on a GPU cluster hosted by Compute Canada which I don't have access anymore.

To answer your question:
1) NaN loss is not a known issue nor have I encountered it before.
2) I don't think a loss of 0.3 indicate a divergence but I agree something might be wrong.

I have not ran the code for a while, to help diagnosing the problem:
1) What kind of GPU and host OS you are using?
2) Did you try to disable the use of tf32 floating point? You can search `allow_tf32` and set those to `False`.

**Comment by fengxiaocurious** [comment]
Thank you very much for your reply. The problem has been resolved, and I forgot to provide feedback in a timely manner.

Previously, I tried to reproduce the results using Torch 2.5 on an NVIDIA 4060 Ti (16GB) with Ubuntu 24, but the reproduction was not successful.

Subsequently, I switched to Torch 2.1 (still on the 4060 Ti 16GB with Ubuntu 24) and applied some stability compensation to the loss function:

```
total_var = torch.exp(variance) + torch.square(labels_err) + 1e-6
wrapper_output = 0.5 * (
    (torch.square(y_true - y_pred) / total_var) + torch.log(total_var + 1e-8)
)
losses = wrapper_output.sum() / y_true.shape[0]
return losses
```
The code is from your init committee. The results now fully match those in the paper. Perhaps the issue was that the code's default configuration `allow_tf32 = True` was being used.

Thanks again for your reply!

**Comment by henrysky** [comment]
Good to know, thanks! I will close the issue.
